### DISCUSSION TITLE
tutorial-extras/ws2812

### DISCUSSION THREAD

**Discussion by giscus** [Air-duino/document#19] [comment]
# tutorial-extras/ws2812

简介 本章介绍使用Air001开发板驱动 WS2812 灯。 WS2812 是一个三色 LED 驱动芯片 (http://world-semi.com/ws2812-family/)，一般集成在 LED 灯珠中。 它仅需一个 IO 传输数据，可通过多个 LED 串联传递信号，实现批量控制。 硬件准备 按☁️ Air001开发板入门 (/tutorial...

https://arduino.luatos.com/tutorial-extras/ws2812.html

**Comment by M0Yi** [comment]
这个就很棒棒的，亏我还找库 原来这就有解决方案！

**Comment by KevinLi-167** [comment]
/*
2024年2月22日13:52:46
spi2的ws2812:
SPI2_MOSI只有NRST和PA4
尝试成功：将SPI2_MOSI的PA4驱动ws2812

*/
//SPI2
#define MOSI_PIN PA4  //PA4 接ws2812
#define MISO_PIN PF1  //PF1
#define SCLK_PIN PB2  //PF0 PA0 PB2 都行

#include <SPI.h>

//https://arduino.luatos.com/library/SPI_2.html#%E4%BE%8B%E5%AD%90
// SPIClass::SPIClass(uint8_t mosi, uint8_t miso, uint8_t sclk, uint8_t ssel)
//            MOSI  MISO  SCLK   官方例子
//新建spi2实例，不使用默认spi PA7，指定引脚为PA4
// SPIClass SPI_2(PA4, PF1, PB2);


//LED灯的个数
#define LED_N 8
//用于存储当前LED灯的状态，默认全为(0,0,0)不亮
unsigned char LED_T[LED_N][3] = { 0 };

void setup(void) {

  //直接使用SPI1或SPI2或者SPI指定引脚，都无法使用
  // SPI.setMOSI(PA7);
  // SPI2->begin();//也不行
  // SPI1_Init();//报错，只有'spi_init这个东西

  //初始化SPI
  SPI_2.begin();
  SPI_2.setBitOrder(MSBFIRST);
  SPI_2.setDataMode(SPI_MODE1);
  // 这里需要让SPI处于8MHz
  // 所以芯片要设置16M，SPI配置为2分频：
  // 16/2=8MHz
  SPI_2.setClockDivider(SPI_CLOCK_DIV2);
  delay(10);
  //刷新一下所有灯的状态，函数见文档接下来的内容
  WS2812_refresh();
}
void WS2812_send(unsigned char r, unsigned char g, unsigned char b) {
  unsigned char bits = 24;
  unsigned long value = 0x00000000;
  value = (((unsigned long)g << 16) | ((unsigned long)r << 8) | ((unsigned long)b));
  while (bits > 0) {
    if ((value & 0x800000) != LOW) {
      SPI_2.transfer(0xF8);  //1
      asm("nop");
      asm("nop");
    } else {
      SPI_2.transfer(0xC0);  //0
    }
    value <<= 1;
    bits--;
  }
}
void WS2812_refresh() {
  unsigned int n = 0;
  for (n = 0; n < LED_N; n++) {
    WS2812_send(LED_T[n][0], LED_T[n][1], LED_T[n][2]);
  }
  delayMicroseconds(60);
}
int count = 0;
void loop(void) {
  count++;
  //把上一次亮的灯灭了
  LED_T[(count - 1) % LED_N][0] = 0;
  LED_T[(count - 1) % LED_N][1] = 0;
  LED_T[(count - 1) % LED_N][2] = 0;
  //点亮这次的这个灯
  //来个浅蓝色吧
  //亮度低一点，不然刺眼
  LED_T[count % LED_N][0] = 5;   //R
  LED_T[count % LED_N][1] = 5;   //G
  LED_T[count % LED_N][2] = 20;  //B
  //刷新所有灯的状态
  WS2812_refresh();
  //延时一小会
  delay(10);

  // if (SPI2 != NULL) { delay(1000); }// 此处SPI2还未去了解是什么
}

**Comment by createskyblue** [comment]
献上一个用NRST引脚点灯的代码
```


#define LED_DATA_PIN PA6
#define BEEP_PIN PA4
#define SW_PIN PA0

#define BIT0_PULSE 0x80
#define BIT1_PULSE 0xF0

SPI_HandleTypeDef Spi2Handle;
DMA_HandleTypeDef DmaHandle;

#define LED_NUM 3
uint8_t LED_Buf[LED_NUM * 24];
void setLED(uint8_t num, uint8_t r, uint8_t g, uint8_t b) {
	uint8_t *p = LED_Buf + num * 24;
	for (int i = 0; i < 8; i++) {
		*p++ = (r & 0x80) ? BIT1_PULSE : BIT0_PULSE;
		r <<= 1;
	}
	for (int i = 0; i < 8; i++) {
		*p++ = (g & 0x80) ? BIT1_PULSE : BIT0_PULSE;
		g <<= 1;
	}
	for (int i = 0; i < 8; i++) {
		*p++ = (b & 0x80) ? BIT1_PULSE : BIT0_PULSE;
		b <<= 1;
	}
}

void updateLED() {
	HAL_SPI_Transmit_DMA(&Spi2Handle, LED_Buf, LED_NUM * 24);
}


uint8_t colorTable[8][3] = {
	{255, 0, 0},
	{0, 255, 0},
	{0, 0, 255},
	{255, 255, 0},
	{255, 0, 255},
	{0, 255, 255},
	{255, 255, 255}
};



extern "C"
{
	void DMA1_Channel1_IRQHandler()
	{
		HAL_DMA_IRQHandler(&DmaHandle);
	}


	/**
	 * @brief This function handles SPI1 Interrupt .
	 */
	void SPI2_IRQHandler()
	{
	HAL_SPI_IRQHandler(&Spi2Handle);
	}

}

/**
  * @brief  写option空间函数
  * @param  无
  * @retval 无
  */
void APP_FlashOBProgram(void)
{
  FLASH_OBProgramInitTypeDef OBInitCfg;

  HAL_FLASH_Unlock();        /* 解锁FLASH */
  HAL_FLASH_OB_Unlock();     /* 解锁OPTION */

  /* 配置OPTION选项 */
  OBInitCfg.OptionType = OPTIONBYTE_USER;
  OBInitCfg.USERType = OB_USER_NRST_MODE;

  OBInitCfg.USERConfig = OB_RESET_MODE_GPIO;

  /* 启动option byte编程 */
  HAL_FLASH_OBProgram(&OBInitCfg);

  HAL_FLASH_Lock();      /* 锁定FLASH */
  HAL_FLASH_OB_Lock();   /* 锁定OPTION */

  /* 产生一个复位，option byte装载 */
  HAL_FLASH_OB_Launch();
}


void setup() {
	__HAL_RCC_SYSCFG_CLK_ENABLE();                  /* 使能SYSCFG时钟 */
	__HAL_RCC_GPIOF_CLK_ENABLE();
	__HAL_RCC_SPI2_CLK_ENABLE();                    /* SPI1时钟使能 */
    __HAL_RCC_DMA_CLK_ENABLE();                     /* DMA时钟使能 */


	Serial.begin(9600);

	/* 判断RST管脚 */
	if(READ_BIT(FLASH->OPTR, FLASH_OPTR_NRST_MODE) == OB_RESET_MODE_RESET)
	{
		/* 写OPTION */
		APP_FlashOBProgram();
	}

	//简洁初始化GPIO
	pinMode(BEEP_PIN, OUTPUT);
	pinMode(SW_PIN, INPUT_PULLUP);

	//复杂初始化GPIO
	GPIO_InitTypeDef GPIO_InitStruct;
	GPIO_InitStruct.Mode = GPIO_MODE_AF_PP;        /* 复用模式 */
	GPIO_InitStruct.Pull = GPIO_PULLDOWN;
	GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
	GPIO_InitStruct.Pin = GPIO_PIN_2;
	GPIO_InitStruct.Alternate = GPIO_AF3_SPI2;
	HAL_GPIO_Init(GPIOF, &GPIO_InitStruct);

	/*中断配置*/
    HAL_NVIC_SetPriority(SPI2_IRQn, 1, 0);
    HAL_NVIC_EnableIRQ(SPI2_IRQn);
	
	Spi2Handle.Instance               = SPI2;                       /* SPI1 */
	Spi2Handle.Init.BaudRatePrescaler = SPI_BAUDRATEPRESCALER_4;    /* 4分频 */
	Spi2Handle.Init.Direction         = SPI_DIRECTION_1LINE;       /* 全双工 */
	Spi2Handle.Init.CLKPolarity       = SPI_POLARITY_LOW;           /* 时钟极性低 */
	Spi2Handle.Init.CLKPhase          = SPI_PHASE_1EDGE ;           /* 数据采样从第一个时钟边沿开始 */
	Spi2Handle.Init.DataSize          = SPI_DATASIZE_8BIT;          /* SPI数据长度为8bit */
	Spi2Handle.Init.FirstBit          = SPI_FIRSTBIT_MSB;           /* 先发送MSB */
	Spi2Handle.Init.NSS               = SPI_NSS_HARD_OUTPUT;        /* NSS软件模式(硬件模式) */
	Spi2Handle.Init.Mode = SPI_MODE_MASTER;                         /* 配置为主机 */
	/*反初始化SPI配置*/
	if (HAL_SPI_DeInit(&Spi2Handle) != HAL_OK)
	{
	Error_Handler();
	}

	/*SPI初始化*/
	if (HAL_SPI_Init(&Spi2Handle) != HAL_OK)
	{
	Error_Handler();
	}

	DmaHandle.Init.Direction = DMA_MEMORY_TO_PERIPH;          /* M2P 模式 */
	DmaHandle.Init.PeriphInc = DMA_PINC_DISABLE;               /* 外设地址增量模式使能 */
	DmaHandle.Init.MemInc = DMA_MINC_ENABLE;                  /* 存储器地址增量模式使能 */
	DmaHandle.Init.PeriphDataAlignment = DMA_PDATAALIGN_BYTE; /* 外设数据宽度为8位 */
	DmaHandle.Init.MemDataAlignment = DMA_MDATAALIGN_BYTE;    /* 存储器数据宽度为8位 */
	DmaHandle.Init.Mode = DMA_NORMAL;                         /* DMA循环模式开启 */
	DmaHandle.Init.Priority = DMA_PRIORITY_VERY_HIGH;              /* 通道优先级为高 */
	DmaHandle.Instance = DMA1_Channel1;                       /* 选择DMA通道1 */
	/* DMA初始化 */
	if (HAL_DMA_Init(&DmaHandle) != HAL_OK)                   
	{
	Error_Handler();
	}

	/*DMA句柄关联到SPI句柄*/
	Spi2Handle.hdmatx = &DmaHandle;
	DmaHandle.Parent = &Spi2Handle;
	HAL_DMA_ChannelMap(&DmaHandle, DMA_CHANNEL_MAP_SPI2_TX);

	HAL_NVIC_SetPriority(DMA1_Channel1_IRQn, 1, 0);
    HAL_NVIC_EnableIRQ(DMA1_Channel1_IRQn);
}

uint8_t color = 0;
void loop() {
	setLED(0, colorTable[color][0], colorTable[color][1], colorTable[color][2]);
	color = (color + 1) % 7;
	updateLED();
	delay(200);
}

```

**Reply by createskyblue** [reply]
48Mhz全速的代码

**Reply by createskyblue** [reply]
下面也是48Mhz，不过用 TIM定时器+DMA的传统办法
···


#define LED_DATA_PIN PA6
#define BEEP_PIN PA4
#define SW_PIN PA0

#define BIT0_PULSE 1
#define BIT1_PULSE 4


DMA_HandleTypeDef DmaHandle;
TIM_HandleTypeDef TimHandle;
TIM_OC_InitTypeDef sConfig;

#define LED_RESET_TIME 75
#define LED_NUMS 3
uint8_t LED_Buf[LED_RESET_TIME + 24 * LED_NUMS];

void setLED(uint8_t num, uint8_t r, uint8_t g, uint8_t b) {
	uint8_t *p = LED_Buf + LED_RESET_TIME + num * 24;
	for (int i = 0; i < 8; i++) {
		*p++ = (r & 0x80) ? BIT1_PULSE : BIT0_PULSE;
		r <<= 1;
	}
	for (int i = 0; i < 8; i++) {
		*p++ = (g & 0x80) ? BIT1_PULSE : BIT0_PULSE;
		g <<= 1;
	}
	for (int i = 0; i < 8; i++) {
		*p++ = (b & 0x80) ? BIT1_PULSE : BIT0_PULSE;
		b <<= 1;
	}
}

void updateLED() {
	HAL_TIM_PWM_Start_DMA(&TimHandle, TIM_CHANNEL_1, (uint32_t*)LED_Buf, sizeof(LED_Buf));
	delay(1);
	HAL_TIM_PWM_Stop_DMA(&TimHandle, TIM_CHANNEL_1);
}

uint8_t colorTable[8][3] = {
	{255, 0, 0},
	{0, 255, 0},
	{0, 0, 255},
	{255, 255, 0},
	{255, 0, 255},
	{0, 255, 255},
	{255, 255, 255}
};


void setup() {
	memset(LED_Buf, 0, sizeof(LED_Buf));

	Serial.begin(9600);

	//简洁初始化GPIO
	pinMode(BEEP_PIN, OUTPUT);
	pinMode(SW_PIN, INPUT_PULLUP);

	//复杂初始化GPIO
	GPIO_InitTypeDef GPIO_InitStruct;
	/* TIM1时钟使能 */
	__HAL_RCC_TIM16_CLK_ENABLE();
	/* GPIOA时钟使能 */
	__HAL_RCC_GPIOA_CLK_ENABLE();

	GPIO_InitStruct.Mode = GPIO_MODE_AF_PP;        /* 复用模式 */
	GPIO_InitStruct.Pull = GPIO_PULLUP;            /* 引脚上拉 */
	GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_VERY_HIGH;

	/*GPIOA6初始化*/
	GPIO_InitStruct.Pin = GPIO_PIN_6;
	GPIO_InitStruct.Alternate = GPIO_AF5_TIM16;
	HAL_GPIO_Init(GPIOA, &GPIO_InitStruct);

	//初始化定时器
 	TimHandle.Instance = TIM16;                                                  /* 选择TIM1 */
	TimHandle.Init.Prescaler = 8;                                 /* 预分频为800-1 */
	TimHandle.Init.Period = 6;                                      /* 自动重装载值 */
	TimHandle.Init.ClockDivision = TIM_CLOCKDIVISION_DIV1;                  /* 时钟不分频 */
	TimHandle.Init.CounterMode = TIM_COUNTERMODE_UP;                      /* 向上计数 */
	TimHandle.Init.RepetitionCounter = 1 - 1;                                   /* 不重复计数 */
	TimHandle.Init.AutoReloadPreload = TIM_AUTORELOAD_PRELOAD_DISABLE;          /* 自动重装载寄存器没有缓冲 */
	/* 基础时钟初始化 */
	if (HAL_TIM_PWM_Init(&TimHandle) != HAL_OK)
	{
		Error_Handler();
	}

	sConfig.OCMode = TIM_OCMODE_PWM1;                                     /* 输出配置为PWM1 */
	sConfig.OCPolarity = TIM_OCPOLARITY_HIGH;                                 /* OC通道输出高电平有效 */
	sConfig.OCFastMode = TIM_OCFAST_DISABLE;                                  /* 输出快速使能关闭 */
	sConfig.OCNPolarity = TIM_OCNPOLARITY_HIGH;                                /* OCN通道输出高电平有效 */
	sConfig.OCNIdleState = TIM_OCNIDLESTATE_RESET;                              /* 空闲状态OC1N输出低电平 */
	sConfig.OCIdleState = TIM_OCIDLESTATE_RESET;                               /* 空闲状态OC1输出低电平 */

	sConfig.Pulse = 0;	//0:2 1:7
	/* 配置通道1 */
	if (HAL_TIM_PWM_ConfigChannel(&TimHandle, &sConfig, TIM_CHANNEL_1) != HAL_OK)
	{
		Error_Handler();
	}

	__HAL_RCC_DMA_CLK_ENABLE();                               /*使能DMA时钟*/
	DmaHandle.Init.Direction = DMA_MEMORY_TO_PERIPH;          /* M2P 模式 */
	DmaHandle.Init.PeriphInc = DMA_PINC_DISABLE;               /* 外设地址增量模式使能 */
	DmaHandle.Init.MemInc = DMA_MINC_ENABLE;                  /* 存储器地址增量模式使能 */
	DmaHandle.Init.PeriphDataAlignment = DMA_PDATAALIGN_HALFWORD; /* 外设数据宽度为16位 */
	DmaHandle.Init.MemDataAlignment = DMA_MDATAALIGN_BYTE;    /* 存储器数据宽度为8位 */
	DmaHandle.Init.Mode = DMA_NORMAL;                         /* DMA循环模式开启 */
	DmaHandle.Init.Priority = DMA_PRIORITY_VERY_HIGH;              /* 通道优先级为高 */

	DmaHandle.Instance = DMA1_Channel1;                       /* 选择DMA通道1 */
	/* DMA初始化 */
	if (HAL_DMA_Init(&DmaHandle) != HAL_OK)                   
	{
	Error_Handler();
	}

	TimHandle.hdma[TIM_DMA_ID_CC1] = &DmaHandle;                               /* 关联DMA */
	DmaHandle.Parent = (void*)&TimHandle;                      /* DMA的父对象为TIM1 */
	HAL_DMA_ChannelMap(&DmaHandle, DMA_CHANNEL_MAP_TIM16_CH1); /* DMA1通道1映射到TIM16_CH1 */
	
	
}

uint8_t color = 0;
void loop() {
	Serial.println("Hello, world!");
	
	setLED(0, colorTable[color][0], colorTable[color][1], colorTable[color][2]);
	color = (color + 1) % 7;

	updateLED();
	delay(200);
}
···

**Reply by createskyblue** [reply]
建议还是用这里贴出的spi办法，相比TIM+DMA能节省100来个字节的RAM还有4字节的Flash 😓
